File: ios-sdk/Sources/ConsentSDK/Analytics/ConsentAnalytics.swift
import Foundation

/// Analytics event types emitted by the Consent SDK.
public enum ConsentEvent {
    case consentShown(language: String)
    case consentSubmitted(selectedPurposes: [String], age: Int, language: String)
    case consentDeclined(language: String)
    case consentDismissed
    case languageChanged(from: String, to: String)
    case error(message: String, code: String)
}

/// Protocol for receiving consent analytics events.
public protocol ConsentEventListener: AnyObject {
    func onEvent(_ event: ConsentEvent)
}

/// Manages analytics event listeners and dispatches events.
public class ConsentAnalytics {
    private var listeners: [ConsentEventListener] = []
    private let lock = NSLock()

    public func addListener(_ listener: ConsentEventListener) {
        lock.lock()
        defer { lock.unlock() }
        listeners.append(listener)
    }

    public func removeListener(_ listener: ConsentEventListener) {
        lock.lock()
        defer { lock.unlock() }
        listeners.removeAll { $0 === listener }
    }

    public func clearListeners() {
        lock.lock()
        defer { lock.unlock() }
        listeners.removeAll()
    }

    internal func dispatch(_ event: ConsentEvent) {
        let snapshot: [ConsentEventListener]
        lock.lock()
        snapshot = listeners
        lock.unlock()
        snapshot.forEach { $0.onEvent(event) }
    }
}

File: core/src/commonMain/kotlin/com/niceoneconsent/sdk/analytics/ConsentAnalytics.kt
package com.niceoneconsent.sdk.analytics

/**
 * Event types emitted by the Consent SDK for analytics tracking.
 */
sealed class ConsentEvent {
    /** Popup was displayed to the user. */
    data class ConsentShown(val language: String) : ConsentEvent()

    /** User successfully submitted their consent. */
    data class ConsentSubmitted(
        val selectedPurposes: List<String>,
        val age: Int,
        val language: String
    ) : ConsentEvent()

    /** User declined consent. */
    data class ConsentDeclined(val language: String) : ConsentEvent()

    /** An error occurred during an SDK operation. */
    data class Error(val message: String, val code: String) : ConsentEvent()

    /** Language was switched. */
    data class LanguageChanged(val from: String, val to: String) : ConsentEvent()

    /** Popup was dismissed. */
    data object ConsentDismissed : ConsentEvent()
}

/**
 * Listener interface for receiving consent analytics events.
 *
 * Client apps implement this to track consent interactions.
 */
fun interface ConsentEventListener {
    fun onEvent(event: ConsentEvent)
}

/**
 * Manages analytics event listeners and dispatches events.
 */
class ConsentAnalytics {
    private val listeners = mutableListOf<ConsentEventListener>()

    /** Register an event listener. */
    fun addListener(listener: ConsentEventListener) {
        listeners.add(listener)
    }

    /** Remove a previously registered listener. */
    fun removeListener(listener: ConsentEventListener) {
        listeners.remove(listener)
    }

    /** Remove all listeners. */
    fun clearListeners() {
        listeners.clear()
    }

    /** Dispatch an event to all registered listeners. */
    internal fun dispatch(event: ConsentEvent) {
        val snapshot = listeners.toList()
        snapshot.forEach { it.onEvent(event) }
    }
}
